### ISSUE TITLE
Simple range over an always nil slice freaks out golangci-lint.

### ISSUE THREAD

**Issue by puellanivis** [golangci/golangci-lint#510]
The following code claims a `nilness: nil dereference in index operation` among a large amount of text that boils down to “I cannot tell you where this is happening”.

```go
package main

func main() {
	var s []string

	for _, v := range s {
		_ = v
	}
}
```

It seems like there is an implicit array index in the `ssa` due to the `range`, and thus the `instr.Pos()` for it returns a `token.NoPos` (as per https://github.com/golang/tools/blob/master/go/ssa/ssa.go#L217-L218 ), which severely breaks all attempts at filtering results, and even with reporting the location of the problem.

Adding `{}` to the end of the `[]string` resolves the message, but finding where the problem is occurring (in my case) within 1425 lines of code spread across 23 files is impossible, especially when one has no idea what one is looking for.

It might be a good idea to account for the possibility of `.Pos() == NoPos` and either disregard the message (as it is unlikely that implicit instructions are going to be performing a nil-deref), or at least using the `fn.Pos()` with some sort of addition to the message like “in some implicit code somewhere in this function”.

Please include the following information:

1. Version of golangci-lint:
```console
$ golangci-test --version
Error: unknown flag: --version
$ git rev-parse HEAD
39f46be46090828834b920f59994c97767694238
```

2. Config file: `cat .golangci.yml`: (empty)

3. Go environment: `go version && go env`
```console
$ go version && go env
go version go1.12.2 linux/amd64
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/snowgirl/.cache/go-build"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPATH="/home/snowgirl/Work/go"
GOPROXY=""
GORACE=""
GOROOT="/usr/lib64/go/1.12"
GOTMPDIR=""
GOTOOLDIR="/usr/lib64/go/1.12/pkg/tool/linux_amd64"
GCCGO="gccgo"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build170324728=/tmp/go-build -gno-record-gcc-switches"
```

4. Verbose output of running: `golangci-lint run -v`
```console
$ golangci-lint run -v
INFO [config_reader] Config search paths: [./ /home/snowgirl/Work/tmp/golangci-proof-of-errors/nil-array-range /home/snowgirl/Work/tmp/golangci-proof-of-errors /home/snowgirl/Work/tmp /home/snowgirl/Work /home/snowgirl /home /] 
INFO [lintersdb] Active 10 linters: [deadcode errcheck gosimple govet ineffassign staticcheck structcheck typecheck unused varcheck] 
INFO [lintersdb] Optimized sublinters [staticcheck gosimple unused] into metalinter megacheck 
INFO [loader] Go packages loading at mode load deps types and syntax took 897.630208ms 
INFO [loader] SSA repr building timing: packages building 12.089994ms, total 70.083259ms 
INFO [runner] worker.1 took 160.673µs with stages: errcheck: 119.417µs, varcheck: 10.022µs, deadcode: 7.298µs, structcheck: 6.123µs, typecheck: 1.077µs 
INFO [runner] worker.3 took 208.556µs with stages: ineffassign: 189.958µs 
INFO [runner] worker.2 took 108.775823ms with stages: govet: 108.76993ms 
WARN [runner] Can't process result by autogenerated_exclude processor: can't filter issue result.Issue{FromLinter:"govet", Text:"nilness: nil dereference in index operation", Pos:token.Position{Filename:"", Offset:0, Line:0, Column:0}, LineRange:(*result.Range)(nil), HunkPos:0, SourceLines:[]string(nil), Replacement:(*result.Replacement)(nil)}: no file path for issue 
WARN [runner] Can't process result by nolint processor: can't filter issue result.Issue{FromLinter:"govet", Text:"nilness: nil dereference in index operation", Pos:token.Position{Filename:"", Offset:0, Line:0, Column:0}, LineRange:(*result.Range)(nil), HunkPos:0, SourceLines:[]string(nil), Replacement:(*result.Replacement)(nil)}: no file path for issue 
WARN [runner/source_code] Failed to get line 0 for file : failed to get file  lines cache: can't get file  bytes from cache: can't read file : open : no such file or directory 
:0: nilness: nil dereference in index operation (govet)
INFO [runner] worker.4 took 354.751584ms with stages: megacheck: 354.744267ms 
INFO [runner] Workers idle times: #1: 354.568814ms, #2: 245.929908ms, #3: 354.549367ms 
INFO [runner] processing took 898.7µs with stages: exclude: 468.004µs, identifier_marker: 178.379µs, source_code: 55.547µs, autogenerated_exclude: 49.546µs, nolint: 47.859µs, path_prettifier: 45.278µs, cgo: 30.944µs, skip_dirs: 7.428µs, max_same_issues: 4.14µs, uniq_by_line: 3.665µs, max_from_linter: 1.96µs, path_shortener: 1.661µs, max_per_file_from_linter: 1.63µs, replacement_builder: 1.003µs, diff: 860ns, skip_files: 530ns, exclude-rules: 266ns 
INFO File cache stats: 0 entries of total size 0B 
INFO Memory: 15 samples, avg is 144.8MB, max is 270.4MB 
INFO Execution took 1.385439877s                  
```

**Comment by jirfag** [comment]
thank you!

**Comment by xor-gate** [comment]
I think we have hit the https://github.com/golang/go/issues/31008 error

**Comment by puellanivis** [comment]
Indeed.

**Comment by benjaminjkraft** [comment]
Just ran into a similar issue (also with `nilness`, different code path).  While again ideally `nilness` would include a position (and I'll report it upstream once we figure out where the error actually is), it seems like it would be nice for `golangci-lint` to not crash on this error, since `go/analysis` seems to consider it valid, if suboptimal?  We could report at the start of the first file in the package, or something.

**Comment by ldez** [comment]
I tried to reproduce with the initial snippet:
```go
package main

func main() {
	var s []string

	for _, v := range s {
		_ = v
	}
}
```

```yml
linters-settings:
  govet:
    enable-all: true

linters:
  disable-all: true
  enable:
    - govet
```

```console
$ golangci-lint run
main.go:6:20: nilness: nil dereference in index operation (govet)
        for _, v := range s {
                          ^
```

I don't have any problems.

**Comment by puellanivis** [comment]
> ```
> $ golangci-lint run
> main.go:6:20: nilness: nil dereference in index operation (govet)
>         for _, v := range s {
>                           ^
> ```
> 
> I don't have any problems.

This message `nil dereference in index operation` is itself the erroneous behavior. There is no nil dereference happening here. A range over a nil slice is a perfectly valid and sound piece of code with well-defined behavior.† 

†: https://groups.google.com/g/golang-nuts/c/gVgVLQU1FFE

**Comment by benjaminjkraft** [comment]
To be clear, the issue I ran into was similar, but the exact message was not the same.  We didn't end up tracking down the code that caused it, sadly.  But if you look at the `nilness` source it's clear that there are a bunch of places it can still report an error with `token.NoPos`, which is what happened to us for unknown reasons, so that's why it would be nice to at least report the error and not crash.

**Comment by ldez** [comment]
Have you try to use [nilness](https://github.com/golang/tools/blob/master/go/analysis/passes/nilness/nilness.go) directly with `go vet` ?

**Comment by benjaminjkraft** [comment]
@ldez it was my coworker who got the issue, and they ended up just working around it, so I'm not sure.  But the error message made things reasonably clear:
```level=warning msg="[runner] Can't process result by nolint processor: can't filter issue result.Issue{FromLinter:\"govet\", Text:\"nilness: nil dereference in field selection\", Severity:\"\", SourceLines:[]string(nil), Replacement:(*result.Replacement)(nil), Pkg:(*packages.Package)(0xc0015a26c0), LineRange:(*result.Range)(nil), Pos:token.Position{Filename:\"\", Offset:0, Line:0, Column:0}, HunkPos:0, ExpectNoLint:false, ExpectedNoLintLinter:\"\"}: no file path for issue"
level=warning msg="[runner/source_code] Failed to get line 0 for file : failed to get file  lines cache: can't get file  bytes from cache: can't read file : open : no such file or directory"
```
Specifically, `go vet` is reporting an error with `Pos:token.NoPos`, and then something in golangci-lint is looking at that and trying to read the file with empty-string filename.  I don't know why `go vet` is reporting such an error, and it may be a bug in this specific case, but it doesn't seem like they intend for it to be impossible.

(If it's more useful, I can file a separate issue, apologies if I hijacked this one!)

**Comment by ldez** [comment]
@benjaminjkraft could you open a new issue with a sample to reproduce your problem? thank you.

---
> But it appears my original issue is resolved.

I will close this issue.
